lib/msal-common/src/client/DeviceCodeClient.ts

```diff
@@ -147,62 +147,69 @@ export class DeviceCodeClient extends BaseClient {
         const deviceCodeExpirationTime = TimeUtils.nowSeconds() + deviceCodeResponse.expiresIn;
         const pollingIntervalMilli = deviceCodeResponse.interval * 1000;
 
+        /**
+         * Breaks the polling with specific conditions.
+         * @param request CommonDeviceCodeRequest
+         * @param deviceCodeResponse DeviceCodeResponse
+         */
+        const continuePolling = (): boolean => {
+            if (request.cancel) {
+                this.logger.error("Token request cancelled by setting DeviceCodeRequest.cancel = true");
+                throw ClientAuthError.createDeviceCodeCancelledError();
+            } else if (userSpecifiedTimeout && userSpecifiedTimeout < deviceCodeExpirationTime && TimeUtils.nowSeconds() > userSpecifiedTimeout) {
+                this.logger.error(`User defined timeout for device code polling reached. The timeout was set for ${userSpecifiedTimeout}`);
+                throw ClientAuthError.createUserTimeoutReachedError();
+            } else if (TimeUtils.nowSeconds() > deviceCodeExpirationTime) {
+                if (userSpecifiedTimeout) {
+                    this.logger.verbose(`User specified timeout ignored as the device code has expired before the timeout elapsed. The user specified timeout was set for ${userSpecifiedTimeout}`);
+                }
+                this.logger.error(`Device code expired. Expiration time of device code was ${deviceCodeExpirationTime}`);
+                throw ClientAuthError.createDeviceCodeExpiredError();
+            }
+            return true;
+        };
+
+        /**
+         * Waits for t number of milliseconds
+         * @param t number
+         * @param value T
+         */
+        function delay<T>(t: number, value?: T): Promise<T | void> {
```

Comment:
Can this be defined somewhere more easily sharable?

Reply:
Also, this function and `continuePolling` are defined in different ways. I would propose they are both added as members of this class.

Reply:
+1 to making them class functions. `delay` can also be a utility function instantiated as needed.

Reply:
Moved the `delay` function to the `TimeUtils` utility class and made `continuePolling` a member of the class.